File: src/utils/convertFile.ts
import assert from 'node:assert/strict'
import path from 'node:path'

import httpErrors from 'http-errors'

import { unoserver } from './unoserver.js'

/** Converts file to a new format */
export async function convertFile(
	srcPath: string,
	format: string,
	options?: { filter?: string },
): Promise<{ targetPath: string }> {
	const ext = path.extname(srcPath).toLowerCase()
	assert(
		ext !== '',
		new httpErrors.BadRequest("Can't detect extension for incoming file"),
	)

	let pathWithoutExtension = srcPath.slice(0, srcPath.length - ext.length)

	if (`.${format}` === ext) {
		pathWithoutExtension += '-1'
	}

	const targetPath = `${pathWithoutExtension}.${format}`

	await unoserver.convert(srcPath, targetPath, { filter: options?.filter })

	return { targetPath }
}

File: internal/server/convert_file.go
package server

import (
	"context"
	"errors"
	"fmt"
	"net/http"
	"path/filepath"
	"strings"

	"github.com/lynxtaa/unoserver-web/internal/converter"
	"github.com/lynxtaa/unoserver-web/internal/httperror"
)

// ErrInvalidExtension is returned when a file has no extension
var ErrInvalidExtension = errors.New("extension is empty")

// convertFile converts srcPath into the given format, placing the result next to
// the source file, and returns its path
func (s *Server) convertFile(
	ctx context.Context,
	srcPath string,
	format string,
	opts converter.ConvertOptions,
) (targetPath string, err error) {
	ext := strings.ToLower(filepath.Ext(srcPath))
	if ext == "" {
		return "", httperror.New(
			ErrInvalidExtension,
			"can't detect extension for incoming file",
			http.StatusBadRequest,
		)
	}

	pathWithoutExtension := strings.TrimSuffix(srcPath, ext)

	if sameSrcAndTargetFormat := "."+format == ext; sameSrcAndTargetFormat {
		pathWithoutExtension += "-1"
	}

	targetPath = pathWithoutExtension + "." + format

	if err := s.converter.Convert(ctx, srcPath, targetPath, opts); err != nil {
		return "", fmt.Errorf("conversion failed: %w", err)
	}

	return targetPath, nil
}
